mypy/build.py

```diff
@@ -778,95 +785,188 @@ def is_file(path: str) -> bool:
+        """
+        Add path to import context.
+        Modifies self.paths in case if arguments satisfy import context state
+        """
+        assert path.endswith((os.path.sep,) + tuple(PYTHON_EXTENSIONS))
+
+        if self.type is not None and self.type != type:
+            return None
+
+        # We can add at most one module implementation to paths
+        # But module can have multiple stubs
+        py_path = path.endswith('.py')
+        if self.has_py and py_path:
+            return None
+
+        if type == ModuleType.namespace:
+            ok = self._verify_namespace(path)
+        else:
+            ok = self._verify_module(path)
+
+        if not ok:
+            return None
+
+        if py_path:
+            self.has_py = True
+
+        self.type = type
+        self.paths.append(path)
+
+    def _verify_module(self, path: str) -> bool:
+        # At this point we already know that that it's valid python path
+        # We only need to check file existence
+        if not is_file(path):
             return False
-    return True
+
+        return True
+
+    def _verify_namespace(self, path: str) -> bool:
+        if os.path.isdir(path):
+            files = set(list_dir(path) or [])  # type: Set[str]
+            if not ('__init__.py' in files or '__init__.pyi' in files):
+                return True
+
+        return False
+
+
+def is_pkg_path(path: str) -> bool:
+    return path.endswith(('__init__.py', '__init__.pyi'))
+
+
+def is_module_path(path: str) -> bool:
+    return not is_pkg_path(path) and path.endswith(('.py', '.pyi'))
+
+
+def is_namespace_path(path: str) -> bool:
+    return path.endswith(os.path.sep)
+
+
+class ModuleDiscovery:
+    def __init__(self,
+                 lib_path: Iterable[str],
+                 namespaces_allowed: bool = False) -> None:
+
+        self.lib_path = tuple(os.path.normpath(p) for p in lib_path)  # type: Tuple[str, ...]
+        self.namespaces_allowed = namespaces_allowed
+        self._find_module_cache = {}  # type: Dict[str, List[str]]
+
+    def find_module(self, id: str) -> Optional[str]:
+        paths = self._find_module(id)
+        if paths:
+            return paths[0]
+        else:
+            return None
+
+    def find_modules_recursive(self, module: str) -> List[BuildSource]:
+        """
+        Discover module and all it's children
+        Remove duplicates from discovered paths
+        """
+        hits = set()  # type: Set[str]
+        result = []  # type: List[BuildSource]
+        for src in self._find_modules_recursive(module):
+            if src.module not in hits:
+                hits.add(src.module)
+                result.append(src)
+        return result
+
+    def _find_modules_recursive(self, module: str) -> List[BuildSource]:
+        module_paths = self._find_module(module)
+
+        srcs = []  # type: List[BuildSource]
+        for path in module_paths:
+            if is_module_path(path) or is_pkg_path(path):
```

Comment:
again here (and below) there's a lot of re-checking of the path type that is redundant; `_find_module` already had this information and threw it away. Can't we have `_find_module` throw away less info, return an object instead of just paths, and avoid unnecessary string path checking?